inbm/dispatcher-agent/dispatcher/sota/oras_util.py

```diff
@@ -0,0 +1,150 @@
+    @param umask: file permission mask
+    @raises DispatcherException: any exception
+    """
+    dispatcher_broker.telemetry(f'Package to be fetched from {uri.value}')
+    dispatcher_broker.telemetry(
+        'Checking authenticity of package by checking signature and source')
+
+    if not isinstance(uri, CanonicalUri):
+        raise DispatcherException("Internal error: uri improperly passed to download function")
+
+    """
+    In case of uri.value = https://registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest
+    source = https://registry-rs.internal.ledgepark.intel.com/one-intel-edge
+    registry_server = registry-rs.internal.ledgepark.intel.com
+    image = tiberos
+    image_tag = latest
+    image_full_path = registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest
+    repository_name = one-intel-edge
+    registry_manifest = https://registry-rs.internal.ledgepark.intel.com/v2/one-intel-edge/tiberos/manifest/latest
+    """
+    try:
+        source = uri.value[:-(len(uri.value.split('/')[-1]) + 1)]
+        registry_server = uri.value.split("/")[2]
+        image = os.path.basename(urlsplit(uri.value).path).split(":")[0]
+        image_tag = uri.value.split(":")[-1]
+        repository_name = "/".join(uri.value.split("/")[3:-1])
+        image_full_path = f"{registry_server}/{repository_name}/{image}:{image_tag}"
+        registry_manifest = f"https://{registry_server}/v2/{repository_name}/{image}/manifests/{image_tag}"
+    except IndexError as err:
+        logger.error(f"IndexError occurs with uri {uri.value}: {err}")
+        raise DispatcherException(err)
+
+    logger.debug(f"source: {source}, "
+                 f"registry_server: {registry_server}, "
+                 f"image: {image}, "
+                 f"image_tag: {image_tag}, "
+                 f"repository_name: {repository_name}, "
+                 f"registry_manifest: {registry_manifest}")
+
+    verify_source(source=source, dispatcher_broker=dispatcher_broker)
+    dispatcher_broker.telemetry('Source Verification check passed')
+
+    try:
+        enough_space = is_enough_space_to_download(
+            registry_manifest, repo, password)
+    except DispatcherException as e:
+        raise DispatcherException(e)
+
+    if not enough_space:
+        err_msg = " Insufficient free space available on " + repo.get_repo_path() + \
+                  " for " + str(uri.value)
+        raise DispatcherException(err_msg)
+
+    if password:
+        logger.debug("JWT Token provided.")
+    else:
+        err_msg = " No JWT token. Abort the update. "
+        raise DispatcherException(err_msg)
+
+    msg = f'Fetching software package from {image_full_path}'
+    dispatcher_broker.telemetry(msg)
+
+    # Call oras to pull the image
+    # The password is the JWT token
+    (out, err_run, code) = PseudoShellRunner().run(f"oras pull {image_full_path} -o {repo.get_repo_path()} --password {password}")
```

Comment:
I set `os.environ["ORAS_PASSWORD"] = password` before running `PseudoShellRunner().run(f"oras pull {image_full_path} -o {repo.get_repo_path()} --password $ORAS_PASSWORD")`.
Will run the test on Monday.

Reply:
> I set `os.environ["ORAS_PASSWORD"] = password` before running `PseudoShellRunner().run(f"oras pull {image_full_path} -o {repo.get_repo_path()} --password $ORAS_PASSWORD")`. Will run the test on Monday.

This is a potential issue too; this sets the environment variable in the current process which means that any process we launch afterwards will see the `ORAS_PASSWORD` value. It's better to only set it specifically for the new process. This will require a change to `PseudoShellRunner`

Reply:
Modified `PseudoShellRunner` with below changes. This will require the apparmor access to  `/usr/bin/bash rix` and `/usr/bin/sh rix`

        elif env:
            proc = subprocess.Popen(
                cmd,
                cwd=cwd,
                shell=True,
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                env=env)